### PR TITLE
WIP: Add improved CSS border style props to View

### PR DESCRIPTION
**Pull request by JoshMarler**
@nick-thompson, Would you mind giving this a quick pass to let me know if you think it's going roughly in the direction before I do any refactoring and quirky behaviour fixes. All the basics are working just fine with this approach. Bit of repetition that can be cleaned up but over all I think the approach came out pretty nicely and offers a decent level of customisation without going too mad.

### CODE REVIEW

File: blueprint/core/blueprint_View.cpp [hunk 1]
```diff
@@ -165,47 +165,23 @@ namespace blueprint
...
-            float borderRadius = getResolvedLengthProperty("border-radius", minLength);
-
-            border.addRoundedRectangle(borderBounds, borderRadius);
-            g.setColour(c);
-            g.strokePath(border, juce::PathStrokeType(borderWidth));
-            g.reduceClipRegion(border);
-        }
-
-        if (props.contains("background-color"))
+        else if (props.contains("background-color"))
```
Comment (nick-thompson) [inline blueprint/core/blueprint_View.cpp:180]: I don't think we want the `else` here right? You could have a border-path and a background-color at the same time no?

File: blueprint/core/blueprint_View.cpp [hunk 2]
```diff
@@ -165,47 +165,23 @@ namespace blueprint
...
-                float borderWidth = props.getWithDefault("border-width", 1.0);
-
-                g.setColour(c);
-                g.strokePath(p, juce::PathStrokeType(borderWidth));
-            }
+            auto drawable = juce::Drawable::createFromImageData(
+                    source.toRawUTF8(),
+                    source.getNumBytesAsUTF8());
 
-            g.reduceClipRegion(p);
```
Comment (nick-thompson) [inline blueprint/core/blueprint_View.cpp:181]: I think we still want this `reduceClipRegion` to make sure background fills clip to the border path, right?

File: blueprint/core/blueprint_View.cpp [hunk 3]
```diff
@@ -165,47 +165,23 @@ namespace blueprint
...
+            //TODO: Additional work for opacity or is current setAlpha implementation enough?
```
Comment (nick-thompson) [inline blueprint/core/blueprint_View.cpp:177]: I think people will probably use the `opacity` prop on the regular `View` for this anyway, so don't worry about it here

**Comment by nick-thompson** [comment]
Yep, this looks like a good direction. I think we can probably refactor some of the js to clean up a bit, but "make it run, make it right, make it fast" and all that.

Before we go too hard this way though: I wonder, do you notice any performance implications when rendering, e.g. long lists of things that each have a border? We're definitely doing a bunch of small string concatenation here, which I think is one of duktape's performance warnings.

I'm also going to poke at some browser implementations of these border renderings just to make sure there's not some obvious alternative way of doing this that I'm totally missing. Will try to get to that this weekend

File: blueprint/core/blueprint_View.cpp (hunk 1 above)
Comment (Prosolve) [reply blueprint/core/blueprint_View.cpp:180]: Right so my thinking here was that in the case of `border-svg` we would be painting the background in JS inside the overall border clipping path. Possible we could push that into native somehow. The assumption was that we would intercept the `background-color` prop when building that SVG.

File: blueprint/core/blueprint_View.cpp (hunk 2 above)
Comment (Prosolve) [reply blueprint/core/blueprint_View.cpp:181]: Sort of answered this in the statement about background being handled in SVG clipping path I think.

**Comment by Prosolve** [comment]
>Before we go too hard this way though: I wonder, do you notice any performance implications when rendering, e.g. long lists of things that each have a border? We're definitely doing a bunch of small string concatenation here, which I think is one of duktape's performance warnings.

Yeah I'm not sure. Totally possible. I originally started working on this in C++ as I wasn't too sure on how the performance  with a pure SVG or Canvas approach would be. I can try this today and take a look at how things are performing. It's certainly the cleanest approach as far as I can tell ..... might not be the fastest though.

File: blueprint/core/blueprint_View.cpp (hunk 1 above)
Comment (JoshMarler) [reply blueprint/core/blueprint_View.cpp:180]: Right so my thinking here was that in the case of border-svg we would be painting the background in JS inside the overall border clipping path. Possible we could push that into native somehow. The assumption was that we would intercept the background-color prop when building that SVG.

File: blueprint/core/blueprint_View.cpp (hunk 2 above)
Comment (JoshMarler) [reply blueprint/core/blueprint_View.cpp:181]: Think I've covered this in comment above.

**Comment by JoshMarler** [comment]
>Before we go too hard this way though: I wonder, do you notice any performance implications when rendering, e.g. long lists of things that each have a border? We're definitely doing a bunch of small string concatenation here, which I think is one of duktape's performance warnings.

Yup. I tackled this in C++ originally as I figured if this is happening at the `View` level I should probably go for the quickest thing possible. The SVG approach is arguably the cleanest.....not sure if it will be the fastest. I can try to get some perf tests on this today.

**Comment by JoshMarler** [comment]
@nick-thompson , Based on our recent discussions to move away from an SVG based approach for CSS border props I'm going to close this and will open up a different MR in due course which takes a more performant approach.

**Pull request closed without merging**